File: frameworks/webflux-bench/src/main/java/webfluxbench/grpc/GrpcImpl.java
package webfluxbench.grpc;

import webfluxbench.repository.DbRepository;
import com.google.protobuf.Empty;
import com.google.protobuf.StringValue;
import org.lognet.springboot.grpc.GRpcService;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import webfluxbench.Queries;
import webfluxbench.ReactorWebfluxServiceGrpc;
import webfluxbench.World;
import webfluxbench.WorldList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@GRpcService
public class GrpcImpl extends ReactorWebfluxServiceGrpc.WebfluxServiceImplBase{

    private final DbRepository dbRepository;

    public GrpcImpl(DbRepository dbRepository) {
        this.dbRepository = dbRepository;
    }

    @Override
    public Mono<StringValue> inMemory(Mono<Empty> request) {
        return Mono.just(StringValue.newBuilder().setValue("Hello, World!").build());
    }

    @Override
    public Mono<World> singleQuery(Mono<Empty> request) {
        return randomWorld();
    }

    @Override
    public Mono<WorldList> multiQueries(Mono<Queries> request){
        return getWorlds(request.flux())
                .collect(WorldList::newBuilder, WorldList.Builder::addWorld)
                .map(WorldList.Builder::build);
    }

    @Override
    public Flux<World> serverStream(Mono<Queries> request) {
         return getWorlds(request.flux());
    }

    @Override
    public Flux<World> bidiStream(Flux<Queries> request) {
        return getWorlds(request);
    }

    @Override
    public Mono<World> clientStream(Flux<Queries> request){
        return getWorlds(request).last();
    }

    private Flux<World> getWorlds(Flux<Queries> request) {
        return request
                .map(Queries::getNumber)
                .map(this::getWorldsFromDb)
                .flatMap(Flux::merge);
    }

    private List<Mono<World>> getWorldsFromDb(Integer number) {
        return Stream.generate(this::randomWorld)
                .limit(number)
                .collect(Collectors.toList());
    }

    private Mono<World> randomWorld() {
        return dbRepository.getWorld(randomWorldNumber())
                .map(GrpcImpl::toProto);
    }

    private static int randomWorldNumber() {
        return 1 + ThreadLocalRandom.current().nextInt(10000);
    }

    private static World toProto(webfluxbench.model.World world) {
        return World.newBuilder()
                .setId(world.id)
                .setRandomNumber(world.randomnumber)
                .build();
    }

}

File: frameworks/netcore-bench/netcore-bench/Grpc/GrpcImpl/GrpcImpl.cs
using System.Threading;
using System.Threading.Tasks;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using Netcorebench;
using NetCoreBench.Models;
using World = Netcorebench.World;

namespace NetCoreBench.Grpc.GrpcImpl
{
    public class GrpcImpl : NetCoreBenchService.NetCoreBenchServiceBase
    {
        private readonly IDb _iDb;

        public GrpcImpl(IDb iDb)
        {
            _iDb = iDb;
        }

        public override Task<StringValue> InMemory(Empty request, ServerCallContext context)
        {
            return Task.FromResult(new StringValue {Value = "Hello World"});
        }

        public override Task<World> SingleQuery(Empty request, ServerCallContext context)
        {
            return _iDb.LoadSingleQueryRow()
                .ContinueWith(ToProto, TaskContinuationOptions.OnlyOnRanToCompletion);
        }

        public override Task<WorldList> MultiQueries(Queries request, ServerCallContext context)
        {
            return _iDb.LoadMultipleQueriesRows(request.Number)
                .ContinueWith(ToProtoList, TaskContinuationOptions.OnlyOnRanToCompletion);
        }

        public override async Task ServerStream(Queries request, IServerStreamWriter<World> responseStream,
            ServerCallContext context)
        {
            await _iDb.LoadMultipleQueriesRows(request.Number)
                .ContinueWith(ToProtoList, TaskContinuationOptions.OnlyOnRanToCompletion);
        }

        public override async Task BidiStream(IAsyncStreamReader<Queries> requestStream,
            IServerStreamWriter<World> responseStream, ServerCallContext context)
        {
            while (await requestStream.MoveNext(CancellationToken.None))
                await _iDb.LoadMultipleQueriesRows(requestStream.Current.Number)
                    .ContinueWith(ToProtoList, TaskContinuationOptions.OnlyOnRanToCompletion);
        }

        private static World ToProto(Task<Models.World> row)
        {
            return new World
            {
                Id = row.Result.Id,
                RandomNumber = row.Result.RandomNumber
            };
        }

        private static WorldList ToProtoList(Task<Models.World[]> row)
        {
            var worldList = new WorldList();
            foreach (var world in row.Result)
                worldList.World.Add(new World
                {
                    Id = world.Id,
                    RandomNumber = world.RandomNumber
                });
            return worldList;
        }
    }
}